src/services/entity_builder.js

```diff
@@ -41,47 +35,51 @@ export default class EntityBuilder {
-      'content.idData.timestamp',
-      'content.idData.sequenceNumber'
-    ]);
-    this.ensureTimestampWithinLimit(asset.content.idData.timestamp);
-    validateFieldsConstrainedToSet(asset, ['content', 'assetId']);
-    validateFieldsConstrainedToSet(asset.content, ['idData', 'signature']);
-
-    this.identityManager.validateSignature(asset.content.idData.createdBy, asset.content.signature, asset.content.idData);
+    validateAndCast(asset)
+      .required([
+        'assetId',
+        'content.idData',
+        'content.signature',
+        'content.idData.createdBy',
+        'content.idData.timestamp',
+        'content.idData.sequenceNumber'
+      ])
+      .fieldsConstrainedToSet(['content', 'assetId'])
+      .fieldsConstrainedToSet(['idData', 'signature'], 'content')
+      .validate(['content.idData.timestamp'], this.isTimestampWithinLimit.bind(this),
+        'Timestamp located too far in the future');
+
+    this.identityManager.validateSignature(asset.content.idData.createdBy, asset.content.signature,
+      asset.content.idData);
   }
 
   validateEvent(event) {
-    validatePathsNotEmpty(event, [
-      'eventId',
-      'content.signature',
-      'content.idData',
-      'content.idData.assetId',
-      'content.idData.createdBy',
-      'content.idData.timestamp',
-      'content.idData.dataHash',
-      'content.idData.accessLevel',
-      'content.data'
-    ]);
+    validateAndCast(event)
+      .required([
+        'eventId',
+        'content.signature',
+        'content.idData',
+        'content.idData.assetId',
+        'content.idData.createdBy',
+        'content.idData.timestamp',
+        'content.idData.dataHash',
+        'content.idData.accessLevel',
+        'content.data'
+      ])
+      .validate(['content.idData.timestamp'], this.isTimestampWithinLimit.bind(this),
+        'Timestamp located too far in the future')
+      .fieldsConstrainedToSet(['content', 'eventId'])
+      .fieldsConstrainedToSet(['idData', 'data', 'signature'], 'content')
+      .isNonNegativeInteger(['content.idData.accessLevel', 'content.idData.timestamp']);
+
     this.eventValidators.forEach((validator) => validator.validate(event.content));
-    this.ensureTimestampWithinLimit(event.content.idData.timestamp);
-    validateFieldsConstrainedToSet(event, ['content', 'eventId']);
-    validateFieldsConstrainedToSet(event.content, ['idData', 'data', 'signature']);
-    validateNonNegativeInteger(event.content.idData.accessLevel, `Access level should be a non-negative integer, instead got ${event.content.idData.accessLevel}`);
-    validateNonNegativeInteger(event.content.idData.timestamp, `Timestamp should be a non-negative integer, instead got ${event.content.idData.timestamp}`);
 
-    this.identityManager.validateSignature(event.content.idData.createdBy, event.content.signature, event.content.idData);
+    this.identityManager.validateSignature(event.content.idData.createdBy, event.content.signature,
```

Comment:
Maybe do this validation via a custom validator in `ValidationAndCasting `?

Reply:
Checking signatures is much more complex then usual validations, I prefer having it separate